zeebe/atomix/cluster/src/main/java/io/atomix/cluster/messaging/impl/HeartbeatHandler.java

```diff
@@ -158,32 +161,35 @@ public void handlerAdded(final ChannelHandlerContext ctx) {
     public void channelRead(final ChannelHandlerContext ctx, final Object msg) {
       boolean heartbeatReceived = forwardHeartbeats;
       if (msg instanceof final ProtocolReply reply) {
-        if (outstandingHeartbeats.contains(reply.id())) {
-          // there's no subject in ProtocolReply, so we rely on oustandingHeartbeats to see if it
-          // was a reply to a heartbeat
-          heartbeatReceived = true;
-        }
+        final var isHeartbeat = outstandingHeartbeats.contains(reply.id());
         // remove all heartbeats sent before (and equal) to the last ProtocolReply received.
         // Note that all ProtocolReply messages (not just heartbeat messages) can indicate that
         // the channel is still open and transmitting.
         // If we remove from the set only the last message received, the size of
         // oustandingHeartbeats will increase for every missed
         // heartbeat and the memory will be freed only when the channel is closed.
         outstandingHeartbeats.removeIfLong(id -> id <= reply.id());
```

Comment:
:question: Considering we remove earlier heartbeats, I may later get the reply - it would just be ignored if I understood correctly, right? Cause it wouldn't be able to tell it's a heartbeat.

Reply:
That should be and edge case as the server replies immediately, since the server `HeartbeatHandler` runs before the "MessageDispatcher".

Anyway if it happens, then it's not marked as an heartbeat response, so it may be forwarded to handlers, which will result in a log line saying some thing like:
"Received message with id[2] but no handler registered"

Reply:
Hm, ideally we would not log this, as it's a really unnecessary warning, and we've had customers open support issues for those :smile: 

I get not keeping track of all forever though :thinking: So we would need a way to know the reply is for a heartbeat, right? 

We can merge this for now, can we create a follow up to address this? In case we do run into this edge case, it's not fully undocumented.